backend/pharmacy/services.py

```diff
@@ -0,0 +1,350 @@
+
+    def delete_prescription(self, id):
+        prescription = self.get_prescription_by_id(id)
+        prescription.delete()
+
+    def add_medicine_to_prescription(self, data):
+        try:
+            prescription = self.get_prescription_by_id(data['prescription_id'])
+            medicine = Medicine.objects.get(pk=data['medicine_id'])
+            detail = PrescriptionDetail(
+                prescription=prescription,
+                medicine=medicine,
+                dosage=data['dosage'],
+                frequency=data['frequency'],
+                duration=data['duration'],
+                prescription_notes=data.get('prescription_notes', ''),
+                quantity=data['quantity']
+            )
+            detail.save()
+            return detail
+        except Medicine.DoesNotExist:
+            raise ValueError(_("Không tìm thấy thuốc với ID: {medicine_id}").format(medicine_id=data['medicine_id']))
+        except Exception as e:
+            raise ValueError(_("Lỗi khi thêm thuốc vào đơn: {error}").format(error=str(e)))
+
+    def get_prescription_details(self, prescription_id):
+        prescription = self.get_prescription_by_id(prescription_id)
+        return prescription.prescription_details.all()
+
+    def update_prescription_detail(self, detail_id, data):
+        detail = self.get_prescription_detail_by_id(detail_id)
+        for key, value in data.items():
+            if key != 'detail_id' and value is not None and getattr(detail, key) != value:
+                setattr(detail, key, value)
+        detail.save()
+        return detail
+
+    def delete_prescription_detail(self, detail_id):
+        detail = self.get_prescription_detail_by_id(detail_id)
+        detail.delete()
+
+    def get_prescriptions_by_patient_id(self, patient_id):
+        try:
+            return Prescription.objects.filter(patient_id=patient_id)
+        except Exception as e:
+            raise ValueError(_("Lỗi khi lấy đơn thuốc theo patient_id: {error}").format(error=str(e)))
+
+    def get_prescriptions_by_appointment_id(self, appointment_id):
+        try:
+            return Prescription.objects.filter(appointment_id=appointment_id)
+        except Exception as e:
+            raise ValueError(_("Lỗi khi lấy đơn thuốc theo appointment_id: {error}").format(error=str(e)))
+
+    def get_prescription_by_id(self, id):
+        try:
+            return Prescription.objects.get(pk=id)
+        except Prescription.DoesNotExist:
+            raise Http404(_("Không tìm thấy đơn thuốc với ID: {id}").format(id=str(id)))
+
+    def get_prescription_detail_by_id(self, id):
+        try:
+            return PrescriptionDetail.objects.get(pk=id)
+        except PrescriptionDetail.DoesNotExist:
+            raise Http404(_("Không tìm thấy chi tiết đơn thuốc với ID: {id}").format(id=str(id)))
+
+    def add_new_medicine(self, data):
+        try:
+            if data['insurance_discount_percent'] < 0 or data['insurance_discount_percent'] > 1:
```

Comment:
% sao lại để range từ 0-1 nhỉ? phải là 0-100, còn khi hệ thống tính toán discount thì mới áp công thức sau chớ